### DISCUSSION TITLE
Build Error when deploying Nextjs app

### DISCUSSION THREAD

**Discussion by LaurensKersbergen** [vercel/next.js#24449] [Help] [answered]
Hey! When I'm trying to deploy my Nextjs app (see: https://github.com/LaurensKersbergen/personal-website), to Vercel.

I've got it all working locally, but when I deploy to Vercel, I keep getting the Errors below. It's a Next.js app, styled with Tailwind CSS and deployed to Vercel. I'm using [this simple setup](https://www.notion.so/Static-Blog-Template-a14c02abc8d9422cb35e4334e4ac88b7).

There seems to be something wrong with the Dropbox integration I'm using. But not sure what I'm doing wrong.

Could you maybe give me a hand? Thanks in advance!

08:27:50.235 | > Build error occurred
08:27:50.235 | Error: ENOENT: no such file or directory, scandir '/vercel/workpath0/_content'
08:27:50.235 | at Object.readdirSync (fs.js:1021:3)
08:27:50.235 | at getPostSlugs (/vercel/workpath0/.next/serverless/pages/[slug].js:1945:55)
08:27:50.235 | at getAllPosts (/vercel/workpath0/.next/serverless/pages/[slug].js:1990:17)
08:27:50.236 | at getStaticPaths (/vercel/workpath0/.next/serverless/pages/[slug].js:2647:85)
08:27:50.236 | at buildStaticPaths (/vercel/workpath0/node_modules/next/dist/build/utils.js:16:86)
08:27:50.236 | at Object.isPageStatic (/vercel/workpath0/node_modules/next/dist/build/utils.js:23:555)
08:27:50.236 | at processTicksAndRejections (internal/process/task_queues.js:93:5) {
08:27:50.236 | type: 'Error',
08:27:50.236 | errno: -2,
08:27:50.236 | syscall: 'scandir',
08:27:50.236 | code: 'ENOENT',
08:27:50.236 | path: '/vercel/workpath0/_content'
08:27:50.236 | }
08:27:50.274 | error Command failed with exit code 1.
08:27:50.275 | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
08:27:50.284 | Error: Command "yarn run build" exited with 1
08:27:52.888 | Done with "package.json"

08:27:30.885 | $ node scripts/downloadDropboxContent.js && next build
08:27:31.168 | (node:660) UnhandledPromiseRejectionWarning: DropboxResponseError: Response failed with a 409 code
08:27:31.168 | at /vercel/workpath0/node_modules/dropbox/cjs/src/response.js:34:11
08:27:31.168 | at processTicksAndRejections (internal/process/task_queues.js:93:5)
08:27:31.168 | at async main (/vercel/workpath0/scripts/downloadDropboxContent.js:27:15)
08:27:31.168 | (Use `node --trace-warnings ...` to show where the warning was created)
08:27:31.168 | (node:660) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
08:27:31.168 | (node:660) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**downloadDropboxContent.js:**

```
require("dotenv").config();

const { Dropbox } = require("dropbox");
const fetch = require("cross-fetch");
const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");

async function main() {
  if (!process.env.DROPBOX_TOKEN) {
    console.error("Set DROPBOX_TOKEN env var and retry");
    process.exit(1);
  }

  if (!process.env.DROPBOX_POSTS_FOLDER_PATH) {
    console.error("Set DROPBOX_POSTS_FOLDER_PATH env var and retry");
    process.exit(1);
  }

  const dbx = new Dropbox({
    accessToken: process.env.DROPBOX_TOKEN,
    fetch: fetch,
  });

  // download content from dropbox as a zip file

  const res = await dbx.filesDownloadZip({
    path: process.env.DROPBOX_POSTS_FOLDER_PATH,
  });

  const outPath = path.join(process.cwd(), "dropbox_download.zip");
  fs.writeFileSync(outPath, res.result.fileBinary);

  // extract it to /_content
  execSync(`rm -rf _content`);
  execSync(`rm -rf _dropbox_download`);
  execSync(`rm -rf public/media`);
  execSync(`unzip dropbox_download.zip -d ./_dropbox_download`);
  // prettier-ignore
  execSync(`mv ./_dropbox_download/${path.basename(process.env.DROPBOX_POSTS_FOLDER_PATH)} ./_content`);
  execSync(`mv ./_content/media ./public/media`);
  execSync(`rm -rf _dropbox_download`);
}

main();

```

**package.json:**

```

{
  "name": "personal-blog-template",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "node scripts/downloadDropboxContent.js && next dev",
    "build": "node scripts/downloadDropboxContent.js && next build",
    "start": "node scripts/downloadDropboxContent.js && next start"
  },
  "dependencies": {
    "@tailwindcss/typography": "^0.3.1",
    "autoprefixer": "^10.1.0",
    "cross-fetch": "^3.0.6",
    "date-fns": "^2.16.1",
    "dotenv": "^8.2.0",
    "dropbox": "^8.2.0",
    "gray-matter": "^4.0.2",
    "next": "10.0.4",
    "postcss": "^8.2.2",
    "react": "17.0.1",
    "react-dom": "17.0.1",
    "remark": "^13.0.0",
    "remark-html": "^13.0.1",
    "tailwindcss": "^2.0.2"
  },
  "devDependencies": {
    "@types/node": "^14.14.41",
    "@types/react": "^17.0.0",
    "cypress": "^7.1.0",
    "typescript": "^4.2.4"
  }
}
```

**Comment by bobarros** [comment]
People cannot help without knowing what the script that is causing the problem do.

downloadDropboxContent.js

**Reply by LaurensKersbergen** [reply]
You're right, sorry! Here's the script:

```
require("dotenv").config();

const { Dropbox } = require("dropbox");
const fetch = require("cross-fetch");
const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");

async function main() {
  if (!process.env.DROPBOX_TOKEN) {
    console.error("Set DROPBOX_TOKEN env var and retry");
    process.exit(1);
  }

  if (!process.env.DROPBOX_POSTS_FOLDER_PATH) {
    console.error("Set DROPBOX_POSTS_FOLDER_PATH env var and retry");
    process.exit(1);
  }

  const dbx = new Dropbox({
    accessToken: process.env.DROPBOX_TOKEN,
    fetch: fetch,
  });

  // download content from dropbox as a zip file

  const res = await dbx.filesDownloadZip({
    path: process.env.DROPBOX_POSTS_FOLDER_PATH,
  });

  const outPath = path.join(process.cwd(), "dropbox_download.zip");
  fs.writeFileSync(outPath, res.result.fileBinary);

  // extract it to /_content
  execSync(`rm -rf _content`);
  execSync(`rm -rf _dropbox_download`);
  execSync(`rm -rf public/media`);
  execSync(`unzip dropbox_download.zip -d ./_dropbox_download`);
  // prettier-ignore
  execSync(`mv ./_dropbox_download/${path.basename(process.env.DROPBOX_POSTS_FOLDER_PATH)} ./_content`);
  execSync(`mv ./_content/media ./public/media`);
  execSync(`rm -rf _dropbox_download`);
}

main();

```

**Comment by bobarros** [comment, marked as answer]
No problem, man. 

We know that this is the error: no such file or directory, scandir '/vercel/workpath0/_content'.

So, your function cannot find this directory.

You have a comment: 

```
// extract it to /_content
```

But right after you're deleting this directory.

**My first question:** are you trying to delete a folder that does not exist in your production environment?

After deleting the folder, you're trying to move some content from an unzipped file to this deleted folder. 

**My second question:** Is this folder some result of the unzipping process? Otherwise, it can't possibly work. You need to create the directory first.

You didn't ask, but I would think if dropbox it's the best option for what you're trying to do.

Good luck 🙃

**Reply by LaurensKersbergen** [reply]
Thanks, I guess I'm almost there. In response to your questions: in what line do you see me deleting the folder?

I am trying to download a Dropbox folder, unzip it, and generate /_content as a build step to use all the files as articles, and use it to produce the site pages. 

Sorry for missing the obvious here. 🙃

**Reply by LaurensKersbergen** [reply]
Btw, when I remove this from .gitignore, it deploys without error:

```
# files generated at build time
dropbox_download.zip
/_content
/public/media
/public/rss.xml
```

Not sure if this is the right approach/hack though?

**Reply by bobarros** [reply]
This line is using the terminal to delete the folder _content:
```
execSync(`rm -rf _content`)
```

In these two, your code tries to move some content to this folder that doesn't exist anymore:
```
execSync(`mv ./_dropbox_download/${path.basename(process.env.DROPBOX_POSTS_FOLDER_PATH)} ./_content`);
execSync(`mv ./_content/media ./public/media`);
```

If you remove these files/directories from .gitignore, they will be uploaded to your repo and Vercel. That's why you don't have any errors anymore. Because now, the folder /content exists. The problem's that you don't know if your site is up to date concerning your dropbox files.

No need to say sorry. That's no such thing as obvious when you are learning something new.

**Reply by LaurensKersbergen** [reply]
Thanks, so if I understand correctly, removing 'execSync(`rm -rf _content`)' solves the issue?

**Reply by bobarros** [reply]
I'm going to give you a solution.

My suggestions are: 

1. Stop using execSync in any of your apps.
2.  Stop using Dropbox for this. 

Before using the code below, you have to change your "blog" folder from dropbox to "Blog"

_____________________

Install adm-zip

```
yarn add adm-zip
or
npm install adm-zip
```

And then, use this function below as your downloadDropboxContent.js.

```javascript

require("dotenv").config();
const { Dropbox } = require("dropbox");
const fetch = require("cross-fetch");
const fs = require("fs");
const path = require("path");
const AdmZip = require('adm-zip');

async function main() {
  if (!process.env.DROPBOX_TOKEN) {
    console.error("Set DROPBOX_TOKEN env var and retry");
    process.exit(1);
  }

  if (!process.env.DROPBOX_POSTS_FOLDER_PATH) {
    console.error("Set DROPBOX_POSTS_FOLDER_PATH env var and retry");
    process.exit(1);
  }

  // sets config to fetch data

  const dbx = new Dropbox({
    accessToken: process.env.DROPBOX_TOKEN,
    fetch: fetch,
  });

  // download content from dropbox as a zip file

  const res = await dbx.filesDownloadZip({
    path: process.env.DROPBOX_POSTS_FOLDER_PATH,
  });

  const outPath = path.join(process.cwd(), "dropbox_download.zip");
  fs.writeFileSync(outPath, res.result.fileBinary);

  // extract data

  var zip = new AdmZip("./dropbox_download.zip");
  await zip.extractAllTo("./_dropbox_download", true);

  // Moves avatar and posts to content folder (creates one if doesnt exist)

  await fs.rename(`_dropbox_download/${path.basename(process.env.DROPBOX_POSTS_FOLDER_PATH)}`, `./_content`, () => {
    console.log("Moved to content");
  })

  // Moves your avatar file to public diretory

  await fs.rename(`./_content/media`, `./public/media`, () => {
    console.log("Moved to public/media");
  })

  // deleting downloaded folder

  try {
    await fs.rmdir("./_dropbox_download", () => {
      console.log("folder _dropbox_download deleted")
    })
  } catch (err) {
    console.error(err)
  }

  // deleting downloaded file

  try {
    await fs.unlink("./dropbox_download.zip", () => {
      console.log("file dropbox_download.zip deleted")
    })
  } catch (err) {
    console.error(err)
  }
}

main();

```

Best of luck.

**Reply by LaurensKersbergen** [reply]
Awesome, that worked! Thanks a lot, man. Learned something new today :)

**Reply by bobarros** [reply]
> Awesome, that worked! Thanks a lot, man. Learned something new today :)

Great to know. As the author of the tread, you can pick my comment and check as the solution for your question. It will be pined up, so people will know this question has an answer and it will be easier for others with similar problems to find it.

I wish you a great week.
